### PR TITLE
Return Skip in InterPodAffinity#PreScore under specific conditions

### PR DESCRIPTION
**Pull request by utam0k**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This commit updates the InterPodAffinity PreScore to return a Skip status when the following conditions are met:
1. There are no nodes to score.
2. The incoming pod has no inter-pod affinities && the `IgnorePreferredTermsOfExistingPods` option is enabled.

#### Which issue(s) this PR fixes:

This change is a part of the work for issue #115745

#### Does this PR introduce a user-facing change?

```release-note
The scheduler skips the InterPodAffinity Score plugin when nothing to do with the Pod.
It will affect some metrics values related to the InterPodAffinity Score plugin.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by utam0k** [comment]
It might be beneficial for @sanposhiho to review the issue since he is the author of this issue.

**Review by Vivekgaddigi — APPROVED** [review]

**Review by kerthcet — COMMENTED** [review]
LGTM, leave for @sanposhiho

**Comment by Vivekgaddigi** [comment]
LGTM

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 1]
```diff
@@ -773,9 +781,14 @@ func TestPreferredAffinity(t *testing.T) {
...
 			if !status.IsSuccess() {
+				if status.Code() != test.wantStatus.Code() {
+					t.Errorf("InterPodAffinity#PreScore() returned unexpected status.Code got: %v, want: %v", status.Code(), test.wantStatus.Code())
+				}
+
 				if !strings.Contains(status.Message(), test.wantStatus.Message()) {
-					t.Errorf("unexpected error: %v", status)
+					t.Errorf("InterPodAffinity#PreScore() returned unexpected status.Message got: %v, want: %v", status.Message(), test.wantStatus.Message())
 				}
 			} else {
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:793]: reduce nest caused by `else`. (I know it's not your code though.)

```suggestion
			return
			}
```

**Review by sanposhiho — COMMENTED** [review]
only one nit

**Comment by sanposhiho** [comment]
@utam0k Also, please help to change the release note:

```
The scheduler skips the InterPodAffinity Score plugin when nothing to do with the Pod.
It will affect some metrics values related to the InterPodAffinity Score plugin.
``

(simplify the description + add the note about metric values)

**Comment by utam0k** [comment]
Hi @sanposhiho, thank you for your review. Please take a look at it when you have a moment. Thank you.

**Review by sanposhiho — APPROVED** [review]
/lgtm

thanks

**Comment by sanposhiho** [comment]
@kubernetes/sig-scheduling-leads  PTAL. (instead of `/assign damemi` given https://github.com/kubernetes/kubernetes/pull/116969)

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 2]
```diff
@@ -773,31 +781,36 @@ func TestPreferredAffinity(t *testing.T) {
...
 
-				if !reflect.DeepEqual(test.expectedList, gotList) {
-					t.Errorf("expected:\n\t%+v,\ngot:\n\t%+v", test.expectedList, gotList)
-				}
+			status = p.(framework.ScorePlugin).ScoreExtensions().NormalizeScore(ctx, state, test.pod, gotList)
+			if !status.IsSuccess() {
+				t.Errorf("unexpected error: %v", status)
 			}
 
+			if !reflect.DeepEqual(test.expectedList, gotList) {
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:811]: Sorry, can we use cmp.Diff() instead since we're refactoring. Another suggestion is we can make refactoring another commit for better reviewing since it's independent. What I learned from the community, next time I think :).

Comment (utam0k) [reply pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go]: @kerthcet Sure! I have fixed it.

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 3]
```diff
@@ -773,31 +782,36 @@ func TestPreferredAffinity(t *testing.T) {
...
-				if !reflect.DeepEqual(test.expectedList, gotList) {
-					t.Errorf("expected:\n\t%+v,\ngot:\n\t%+v", test.expectedList, gotList)
-				}
+			status = p.(framework.ScorePlugin).ScoreExtensions().NormalizeScore(ctx, state, test.pod, gotList)
+			if !status.IsSuccess() {
+				t.Errorf("unexpected error: %v", status)
 			}
 
+			if diff := cmp.Diff(test.expectedList, gotList); diff != "" {
+				t.Errorf("node score list dones not match (-want,+got): \n %s", diff)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:813]: typo
```suggestion
				t.Errorf("node score list doesn't match (-want,+got): \n %s", diff)
```

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 4]
```diff
@@ -773,31 +782,36 @@ func TestPreferredAffinity(t *testing.T) {
...
 				}
+				gotList = append(gotList, framework.NodeScore{Name: nodeName, Score: score})
+			}
 
-				if !reflect.DeepEqual(test.expectedList, gotList) {
-					t.Errorf("expected:\n\t%+v,\ngot:\n\t%+v", test.expectedList, gotList)
-				}
+			status = p.(framework.ScorePlugin).ScoreExtensions().NormalizeScore(ctx, state, test.pod, gotList)
+			if !status.IsSuccess() {
+				t.Errorf("unexpected error: %v", status)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:809]:
```suggestion
				t.Errorf("unexpected error from NormalizeScore: %v", status)
```

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 5]
```diff
@@ -773,31 +782,36 @@ func TestPreferredAffinity(t *testing.T) {
...
+				return
+			}
 
-				status = p.(framework.ScorePlugin).ScoreExtensions().NormalizeScore(ctx, state, test.pod, gotList)
+			var gotList framework.NodeScoreList
+			for _, n := range test.nodes {
+				nodeName := n.ObjectMeta.Name
+				score, status := p.(framework.ScorePlugin).Score(ctx, state, test.pod, nodeName)
 				if !status.IsSuccess() {
 					t.Errorf("unexpected error: %v", status)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:802]:
```suggestion
					t.Errorf("unexpected error from Score: %v", status)
```

**Review by sanposhiho — COMMENTED** [review]
Please help other refactor along with that

**Review by sanposhiho — COMMENTED** [review]
Please help other refactor along with that

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go (hunk 3 above)
Comment (utam0k) [reply pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:813]: 🤦

**Comment by utam0k** [comment]
> Please help other refactor along with that

@sanposhiho 
Sure. I too had thought this refactoring you suggested is necessary.

**Review by sanposhiho — APPROVED** [review]
/lgtm

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring.go [hunk 6]
```diff
@@ -145,10 +145,7 @@ func (pl *InterPodAffinity) PreScore(
...
 	// if the incoming pod has no inter-pod affinities.
 	if pl.args.IgnorePreferredTermsOfExistingPods && !hasPreferredAffinityConstraints && !hasPreferredAntiAffinityConstraints {
-		cycleState.Write(preScoreStateKey, &preScoreState{
-			topologyScore: make(map[string]map[string]int64),
-		})
-		return nil
+		return framework.NewStatus(framework.Skip)
 	}
 
 	// Unless the pod being scheduled has preferred affinity terms, we only
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring.go:151]: if `index == 0` below, we could also return Skip.

**Comment by utam0k** [comment]
@alculquicondor Thank you for taking the time to review this. Please take another look 🙏
